lxd/main_forkproxy.go

```diff
@@ -5,129 +5,281 @@ import (
-		return;
-	}
-	listen_pid = atoi(cur);
+	if (cur == NULL ||
+	    (strcmp(cur, "--help") == 0 || strcmp(cur, "--version") == 0 ||
+	     strcmp(cur, "-h") == 0))
+		_exit(EXIT_FAILURE);
 
-	// Get the arguments
-	advance_arg(true);
+	listen_pid = atoi(cur);
+	listen_addr = advance_arg(true);
 	connect_pid = atoi(advance_arg(true));
-	advance_arg(true);
-	fdnum = atoi(advance_arg(true));
-	forked = atoi(advance_arg(true));
-	logPath = advance_arg(true);
-	pidPath = advance_arg(true);
-
-	// Check if proxy daemon already forked
-	if (forked == 0) {
-		logFile = fopen(logPath, "w+");
-		if (logFile == NULL) {
-			_exit(1);
+	connect_addr = advance_arg(true);
+	log_path = advance_arg(true);
+	pid_path = advance_arg(true);
+
+	close(STDIN_FILENO);
+	log_fd = open(log_path, O_WRONLY | O_CREAT | O_CLOEXEC | O_TRUNC, 0600);
+	if (log_fd < 0)
+		_exit(EXIT_FAILURE);
+
+	ret = dup3(log_fd, STDOUT_FILENO, O_CLOEXEC);
+	if (ret < 0)
+		_exit(EXIT_FAILURE);
+
+	ret = dup3(log_fd, STDERR_FILENO, O_CLOEXEC);
+	if (ret < 0)
+		_exit(EXIT_FAILURE);
+
+	pid_file = fopen(pid_path, "w+");
+	if (!pid_file) {
+		fprintf(stderr,
+			"%s - Failed to create pid file for proxy daemon\n",
+			strerror(errno));
+		_exit(EXIT_FAILURE);
+	}
+
+	ret = socketpair(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0, sk_fds);
+	if (ret < 0) {
+		fprintf(stderr,
+			"%s - Failed to create anonymous unix socket pair\n",
+			strerror(errno));
+		_exit(EXIT_FAILURE);
+	}
+	fprintf(stderr, "Created anonymous pair {%d,%d} of unix sockets\n",
+		sk_fds[0], sk_fds[1]);
+
+	pid = fork();
+	if (pid < 0) {
+		fprintf(stderr, "%s - Failed to create new process\n",
+			strerror(errno));
+		_exit(EXIT_FAILURE);
+	}
+
+	if (pid == 0) {
+		whoami = FORKPROXY_CHILD;
+
+		fclose(pid_file);
+		close(sk_fds[0]);
+
+		// Attach to the user namespace of the listener
+		attach_userns(listen_pid);
+
+		// Attach to the network namespace of the listener
+		ret = dosetns(listen_pid, "net");
+		if (ret < 0) {
+			fprintf(stderr, "Failed setns to listener network namespace: %s\n",
+				strerror(errno));
+			_exit(EXIT_FAILURE);
 		}
 
-		if (dup2(fileno(logFile), STDOUT_FILENO) < 0) {
-			fprintf(logFile, "Failed to redirect STDOUT to logfile: %s\n", strerror(errno));
-			_exit(1);
+		// Attach to the mount namespace of the listener
+		ret = dosetns(listen_pid, "mnt");
+		if (ret < 0) {
+			fprintf(stderr, "Failed setns to listener mount namespace: %s\n",
+				strerror(errno));
+			_exit(EXIT_FAILURE);
 		}
-		if (dup2(fileno(logFile), STDERR_FILENO) < 0) {
-			fprintf(logFile, "Failed to redirect STDERR to logfile: %s\n", strerror(errno));
-			_exit(1);
+
+		if (strncmp(listen_addr, "unix:", sizeof("unix:") - 1) == 0) {
+			cur = listen_addr + sizeof("unix:") - 1;
+			if (*cur != '@')
+				if (unlink(cur) < 0 && errno != ENOENT)
+					fprintf(stderr, "%s - Failed to remove old listening socket \"%s\"",
+						strerror(errno), cur);
```

Comment:
So why are we doing this in C when the listener is written in Go?
Wouldn't it be easier to follow if we had the check + unlink done in Go?

Reply:
This needs to be done while being attached to the correct namespaces. In particular the mount namespace that we've created the socket in.

Reply:
Sure but the Go process will still be in those namespaces, it's got to be otherwise it couldn't bind the socket.

Reply:
Right, that was from the previous implementation where this was called twice, I think.